pkg/store/bucket.go

```diff
@@ -308,67 +309,102 @@ type noopGate struct{}
 func (noopGate) Start(context.Context) error { return nil }
 func (noopGate) Done()                       {}
 
+// BucketStoreOption are functions that configure BucketStore.
+type BucketStoreOption func(s *BucketStore)
+
+// WithLogger sets the BucketStore logger to the one you pass.
+func WithLogger(logger log.Logger) BucketStoreOption {
+	return func(s *BucketStore) {
+		s.logger = logger
+	}
+}
+
+// WithRegistry sets a registry that BucketStore uses to register metrics with.
+func WithRegistry(reg prometheus.Registerer) BucketStoreOption {
+	return func(s *BucketStore) {
+		s.reg = reg
+	}
+}
+
+// WithIndexCache sets a indexCache to use instead of a noopCache.
+func WithIndexCache(cache storecache.IndexCache) BucketStoreOption {
+	return func(s *BucketStore) {
+		s.indexCache = cache
+	}
+}
+
+// WithQueryGate sets a queryGate to use instead of a noopGate.
+func WithQueryGate(queryGate gate.Gate) BucketStoreOption {
+	return func(s *BucketStore) {
+		s.queryGate = queryGate
+	}
+}
+
+// WithChunkPool sets a pool.Bytes to use for chunks.
+func WithChunkPool(chunkPool pool.Bytes) BucketStoreOption {
+	return func(s *BucketStore) {
+		s.chunkPool = chunkPool
+	}
+}
+
+// WithFilterConfig sets a filter which Store uses for filtering metrics based on time.
+func WithFilterConfig(filter *FilterConfig) BucketStoreOption {
+	return func(s *BucketStore) {
+		s.filterConfig = filter
+	}
+}
+
+// WithDebugLogging enables debug logging.
+func WithDebugLogging() BucketStoreOption {
+	return func(s *BucketStore) {
+		s.debugLogging = true
+	}
+}
+
 // NewBucketStore creates a new bucket backed store that implements the store API against
 // an object store bucket. It is optimized to work against high latency backends.
-// TODO(bwplotka): Move to config at this point.
 func NewBucketStore(
```

Comment:
What I liked about the previous approach (taking all arguments in input) is that, whenever there's a change here, downstream projects compilation breaks and so it's very loud. With the options approach this is no more guaranteed.

Reply:
Yeah, that's a fair point. On the other hand passing 10x `nil` to the constructor doesn't seem very nice either. I guess - tradeoffs.

Reply:
If passing 10x nil is due by tests maybe we can have a wrapper function `newTestBucketStore()`? Or am I missing anything?

Reply:
I guess that would work too, yes.
So far I've never run into problems with the functional options either and it's pretty commonly used.

Reply:
Fair enough. I don't have a strong opinion. My comment is definitely not a blocker.

Reply:
I see @pracucci point. The sad thing about BucketStore is that it's only used once. Such options are very much more needed for things which are heavily used e.g `block.Upload` function.

I think I am fine with both options here... No strong opinion (:

Reply:
Plus looks like it's done for optional structs so maybe ok? Less work for you @pracucci too?

Reply:
As already mentioned I don't have a strong opinion. I see why you're heading towards functional options and I'm fine too.